### DISCUSSION TITLE
How to call read-only function from other contract address inside the event handler?

### DISCUSSION THREAD

**Discussion by pyk** [ponder-sh/ponder#179] [Question] [answered]
Hey. I'm trying to rewrite the univ2 subgraph with ponder for learning purpose, and currently I'm stuck with how to get the token metadata (name, symbol, etc). 

AFAIK we can't define Contract object without address in ponder, hence I cannot call `name()` for dynamic ERC20 address. 

Given contract address inside the event handler, what is the ponder-way to execute a read-only function of the given address? 

I'm trying to use `viem` to execute the read-only function, but the problem is that I cannot determine which chain and which rpc url the event handler is currently executing on. is there any way to access viem instance inside the event handler?

Thanks!

**Comment by typedarray** [comment, marked as answer]
Thanks for the question! A few points here:

- At the moment, Ponder does not have a pattern for handling projects that use a factory pattern like UniV2. I'm actively thinking through how we should solve for this "dynamic data sources" requirement, which is very common. Ideas welcome!
- The `context.contracts` object currently includes a `ReadOnlyContract` object for each contract defined in `ponder.config.ts`. This has proven convenient for a lot of projects, but it's not as powerful as just having the `viem.PublicClient` object for each network. There's nothing stopping us from including those. It could look like this:
```ts
ponder.on("ContractName:EventName", async ({ event, context }) => {
  const client = context.networks.mainnet.client // viem.PublicClient object

  const response = await client.readContract( /* ... */)
 
  // ...
});
```

The problem with exposing this is that there are some footguns. The `ReadOnlyContract` objects are "safe" because whenever you make a contract call to them, the `eth_call` request is made at the block height of the event log being handled. This allows Ponder to cache the request, which makes hot reloads much faster and saves on RPC usage. If users start making arbitrary RPC requests in their handlers, it could slow down indexing quite a bit and use a lot of RPC credits. So, ideally the solution here would still cache requests where possible. Also, I wouldn't want users doing stuff like `client.getBlockNumber()` when the block number is already available at `event.block.number`.

Will continue to think on this - appreciate any ideas!

**Reply by pyk** [reply]
Good point!

> At the moment, Ponder does not have a pattern for handling projects that use a factory pattern like UniV2. I'm actively thinking through how we should solve for this "dynamic data sources" requirement, which is very common. Ideas welcome!

Not sure if this is a good idea, but I think we can add optional field `address` in `CallOverrides` here:

https://github.com/0xOlias/ponder/blob/f2d88c8462cf94b8e430dfcc9df9d68812525f4f/packages/core/src/handlers/buildInjectedContract.ts#L20C6-L30

Then replace these three lines (including cache key)

https://github.com/0xOlias/ponder/blob/f2d88c8462cf94b8e430dfcc9df9d68812525f4f/packages/core/src/handlers/buildInjectedContract.ts#L69

https://github.com/0xOlias/ponder/blob/f2d88c8462cf94b8e430dfcc9df9d68812525f4f/packages/core/src/handlers/buildInjectedContract.ts#L95

https://github.com/0xOlias/ponder/blob/f2d88c8462cf94b8e430dfcc9df9d68812525f4f/packages/core/src/handlers/buildInjectedContract.ts#L84

with specified address.

So we can do something like the following inside the event handler:

```ts
const { ERC20 } = context.contracts;
const name = await ERC20.name({address: event.params.token0});
```

As a devs, this approach will give me ability to call defined contract with custom/dynamic address with extra features from ponder (typing, safety, caching, etc).
